platform/kube/inject/initializer.go

```diff
@@ -0,0 +1,456 @@
+	// of watched resources for initialization.
+	ResyncPeriod time.Duration
+
+	// Hub is the image registry used for the injected sidecar proxy
+	// and init images, e.g. docker.io/istio.
+	Hub string
+
+	// Tag is the docker image tag used for the injected sidecar proxy
+	// and init images.
+	Tag string
+
+	// Namespace is the Kubernetes namespace the initializer is
+	// responsible for managing. The initializer will manage injection
+	// for all namespaces if the value of v1.NamespaceAll is
+	// specified.
+	Namespace string
+
+	// InjectionPolicy determines the default injection policy for
+	// resources in the managed namespace.
+	InjectionPolicy InjectionPolicy
+}
+
+// Initializer implements a k8s initializer for transparently
+// injecting the sidecar into user resources. For each resource in the
+// managed namespace, the initializer will remove itself from the
+// pending list of initializers can optionally inject the sidecar
+// based on the InjectionPolicy and per-resource policy (see
+// istioSidecarAnnotationPolicyKey).
+type Initializer struct {
+	clientset   kubernetes.Interface
+	mesh        *proxyconfig.ProxyMeshConfig
+	controllers []cache.Controller
+	options     InitializerOptions
+	params      Params
+}
+
+// NewInitializer creates a new instance of the Istio sidecar initializer.
+func NewInitializer(cl kubernetes.Interface, mesh *proxyconfig.ProxyMeshConfig, o InitializerOptions) *Initializer {
+	i := &Initializer{
+		clientset: cl,
+		mesh:      mesh,
+		options:   o,
+		params: Params{
+			InitImage:         InitImageName(o.Hub, o.Tag),
+			ProxyImage:        ProxyImageName(o.Hub, o.Tag),
+			Verbosity:         DefaultVerbosity,
+			SidecarProxyUID:   DefaultSidecarProxyUID,
+			EnableCoreDump:    true,
+			Version:           version.Line(),
+			Mesh:              mesh,
+			MeshConfigMapName: "istio",
+		},
+	}
+
+	kinds := []struct {
+		resource   string
+		getter     cache.Getter
+		objType    runtime.Object
+		initialize func(in interface{}) error
+	}{
+		{
+			"deployments",
+			i.clientset.AppsV1beta1().RESTClient(),
+			&appsv1beta1.Deployment{},
+			i.initializeDeployment,
+		},
+		{
+			"statefulsets",
+			i.clientset.AppsV1beta1().RESTClient(),
+			&appsv1beta1.StatefulSet{},
+			i.initializeStatefulSet,
+		},
+		{
+			"jobs",
+			i.clientset.BatchV1().RESTClient(),
+			&batchv1.Job{},
+			i.initializeJob,
+		},
+		{
+			"daemonsets",
+			i.clientset.ExtensionsV1beta1().RESTClient(),
+			&v1beta1.DaemonSet{},
+			i.initializeDaemonSet,
+		},
+		{
+			"replicasets",
+			i.clientset.ExtensionsV1beta1().RESTClient(),
+			&v1beta1.ReplicaSet{},
+			i.initializeReplicaSet,
+		},
+		{
+			"replicationcontrollers",
+			i.clientset.CoreV1().RESTClient(),
+			&v1.ReplicationController{},
+			i.initializeReplicationController,
+		},
+		{
+			"pods",
+			i.clientset.CoreV1().RESTClient(),
+			&v1.Pod{},
+			i.initializePod,
+		},
+	}
+
+	for _, kind := range kinds {
+		kind := kind // capture the current value of `kind`
```

Comment:
but do we need this trick? can we just do `for _, k := range kinds` so that code is cleaner and easier for others to follow?